includes/class-wc-gateway-stripe.php

```diff
@@ -811,61 +815,45 @@ public function process_payment( $order_id, $retry = true, $force_save_source =
 
 			WC_Stripe_Logger::log( "Info: Begin processing payment for order $order_id for the amount of {$order->get_total()}" );
 
-			$intent = null;
-			if ( $this->stripe_checkout ) {
-				$response = $this->charge_source( $prepared_source, $order, $previous_error );
+			$intent = $this->get_intent_from_order( $order );
+			if ( $intent ) {
+				$intent = $this->update_existing_intent( $intent, $order, $prepared_method );
 			} else {
-				$intent = $this->get_intent_from_order( $order );
-
-				if ( $intent ) {
-					$intent = $this->update_existing_intent( $intent, $order, $prepared_source );
-				} else {
-					$intent = $this->create_and_confirm_intent( $order, $prepared_source );
-					$response = $intent;
-				}
+				$intent = $this->create_and_confirm_intent( $order, $prepared_method );
 			}
 
-			if ( ! empty( $response->error ) ) {
-				if ( ! $prepared_source->is_intent ) {
-					$this->maybe_remove_non_existent_customer( $response->error, $order );
-				}
-
+			if ( ! empty( $intent->error ) ) {
 				// We want to retry.
-				if ( $this->is_retryable_error( $response->error ) ) {
-					return $this->retry_after_error( $response, $order, $retry, $force_save_source, $previous_error );
+				if ( $this->is_retryable_error( $intent->error ) ) {
+					return $this->retry_after_error( $intent, $order, $retry, $force_save_source, $previous_error );
 				}
 
-				$this->throw_localized_message( $response, $order );
+				$this->throw_localized_message( $intent, $order );
 			}
 
-			if ( ! empty( $intent ) ) {
-				// Use the last charge within the intent to proceed.
-				$response = end( $intent->charges->data );
-
-				// If the intent requires a 3DS flow, redirect to it.
-				if ( 'requires_source_action' === $intent->status ) {
-					if ( is_wc_endpoint_url( 'order-pay' ) ) {
-						$redirect_url = add_query_arg( 'wc-stripe-confirmation', 1, $order->get_checkout_payment_url( false ) );
-					} else {
-						/**
-						 * This URL contains only a hash, which will be sent to `checkout.js` where it will be set like this:
-						 * `window.location = result.redirect`
-						 * Once this redirect is sent to JS, the `onHashChange` function will execute `handleCardPayment`.
-						 */
-
-						// Remove this `else` block to avoid hash-based actions.
-						$redirect_url = '#confirm-pi-' . $intent->client_secret . ':' . rawurlencode( $this->get_return_url( $order ) );
-					}
-
-					return array(
-						'result'   => 'success',
-						'redirect' => $redirect_url,
-					);
+			// If the intent requires a 3DS flow, redirect to it.
+			if ( 'requires_source_action' === $intent->status ) {
```

Comment:
Note that this changes with newer API versions: https://stripe.com/docs/upgrades#2019-02-11

Reply:
https://github.com/woocommerce/woocommerce-gateway-stripe/pull/834 is the reason why this PR can't be automatically merged, and it fixes that.